File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    constructor() {
        this.memo = [];
    }
    find(nums, curr, prev) {
        if(curr < 0) {
            return 0;
        }
        if(this.memo[curr][prev+1] !== 0) {
            return this.memo[curr][prev+1];
        }
        const nottake = this.find(nums, curr-1, prev);
        let take = 0;
        if(prev === -1 || nums[curr] < nums[prev]) {
            take = 1 + this.find(nums, curr-1, curr);
        }
        this.memo[curr][prev+1] = Math.max(nottake, take);
        return this.memo[curr][prev+1];
    }
    lengthOfLIS(nums) {
        const n = nums.length;
        this.memo = new Array(n+1)
                        .fill(new Array(n+1)
                                .fill(0));
        console.log(this.memo);
        return this.find(nums, n-1, -1);
    }
}

File: Data Structures & Algorithms/jump-game-ii/submission-1.cs
public class Solution {
    public int Jump(int[] nums) {
        int n = nums.Length - 1;
        int l = 0, r = 0;
        int count = 0;
        while(r < n) {
            int farthest = nums[r];
            for(int i = l; i <= r; i++) {
                farthest = Math.Max(farthest, nums[i] + i);
            }
            l = r + 1;
            r = farthest;
            count++;
        }
        return count;
    }
}
